source/extensions/io_socket/user_space/io_socket_handle_impl.cc

```diff
@@ -0,0 +1,330 @@
+      // Notify the peer we won't write more data. shutdown(WRITE).
+      peer_handle_->setWriteEnd();
+      // Notify the peer that we no longer accept data. shutdown(RD).
+      peer_handle_->onPeerDestroy();
+      peer_handle_ = nullptr;
+    } else {
+      ENVOY_LOG(trace, "socket {} close after peer closed.", static_cast<void*>(this));
+    }
+  }
+  closed_ = true;
+  return Api::ioCallUint64ResultNoError();
+}
+
+bool IoSocketHandleImpl::isOpen() const { return !closed_; }
+
+Api::IoCallUint64Result IoSocketHandleImpl::readv(uint64_t max_length, Buffer::RawSlice* slices,
+                                                  uint64_t num_slice) {
+  if (!isOpen()) {
+    return {0,
+            // TODO(lambdai): Add EBADF in Network::IoSocketError and adopt it here.
+            Api::IoErrorPtr(new Network::IoSocketError(SOCKET_ERROR_INVAL),
+                            Network::IoSocketError::deleteIoError)};
+  }
+  if (pending_received_data_.length() == 0) {
+    if (receive_data_end_stream_) {
+      return {0, Api::IoErrorPtr(nullptr, [](Api::IoError*) {})};
+    } else {
+      return {0, Api::IoErrorPtr(Network::IoSocketError::getIoSocketEagainInstance(),
+                                 Network::IoSocketError::deleteIoError)};
+    }
+  }
+  uint64_t bytes_offset = 0;
+  for (uint64_t i = 0; i < num_slice && bytes_offset < max_length; i++) {
+    auto bytes_to_read_in_this_slice =
+        std::min(std::min(pending_received_data_.length(), max_length) - bytes_offset,
+                 uint64_t(slices[i].len_));
+    // Copy and drain, so pending_received_data_ always copy from offset 0.
+    pending_received_data_.copyOut(0, bytes_to_read_in_this_slice, slices[i].mem_);
+    pending_received_data_.drain(bytes_to_read_in_this_slice);
+    bytes_offset += bytes_to_read_in_this_slice;
+  }
+  auto bytes_read = bytes_offset;
+  ASSERT(bytes_read <= max_length);
+  ENVOY_LOG(trace, "socket {} readv {} bytes", static_cast<void*>(this), bytes_read);
+  return {bytes_read, Api::IoErrorPtr(nullptr, [](Api::IoError*) {})};
+}
+
+Api::IoCallUint64Result IoSocketHandleImpl::read(Buffer::Instance& buffer,
+                                                 absl::optional<uint64_t> max_length_opt) {
+  const uint64_t max_length = max_length_opt.value_or(UINT64_MAX);
+  if (max_length == 0) {
+    return Api::ioCallUint64ResultNoError();
+  }
+  if (!isOpen()) {
+    return {0, Api::IoErrorPtr(new Network::IoSocketError(SOCKET_ERROR_INVAL),
+                               Network::IoSocketError::deleteIoError)};
+  }
+  if (pending_received_data_.length() == 0) {
+    if (receive_data_end_stream_) {
+      return {0, Api::IoErrorPtr(nullptr, [](Api::IoError*) {})};
+    } else {
+      return {0, Api::IoErrorPtr(Network::IoSocketError::getIoSocketEagainInstance(),
+                                 Network::IoSocketError::deleteIoError)};
+    }
+  }
+  // TODO(lambdai): Move slice by slice until high watermark.
```

Comment:
What are your plans to address this TODO?

PR #14054 which changed the argument to read to max_length_opt also effectively imposed a 128kb max read operation size due to the way reservation works now.  In the case of watermark buffers, it reduces read size further by having reserve overrides that look at highWatermarkTriggered and highWatermark

I would recommend using a more reasonable max read size instead of uint64_max.  Paying attention to highWatermark would also be nice.  Possible computation:

```
const uint64_t max_length = max_length_opt.value_or(128 * 1024);
if (buffer.highWatermark() != 0) {
  if (buffer.length() < buffer.highWatermark()) {
    max_length = std::min(max_length, std::max(16384, buffer.highWatermark() - buffer.length()));
  } else {
    max_length = std::min(max_length, 16384);
  }
}
```

Reply:
The logic to move at most N bytes that I'm suggesting above is the best way to accomplish the 3 goals you state.

Moving slice by slice should be avoided.

Reply:
fixed following https://github.com/envoyproxy/envoy/pull/14917#discussion_r569891243